session/filesync/filesync.go

```diff
@@ -229,47 +233,99 @@ func FSSync(ctx context.Context, c session.Caller, opt FSSendRequestOpt) error {
 	return pr.recvFn(stream, opt.DestDir, opt.CacheUpdater, opt.ProgressCb, opt.Differ, opt.Filter)
 }
 
-// NewFSSyncTargetDir allows writing into a directory
-func NewFSSyncTargetDir(outdir string) session.Attachable {
-	p := &fsSyncTarget{
+type FSSyncTarget interface {
+	target() *fsSyncTarget
+}
+
+type fsSyncTarget struct {
+	id     string
+	outdir string
+	f      FileOutputFunc
+}
+
+func (target *fsSyncTarget) target() *fsSyncTarget {
+	return target
+}
+
+func WithFSSync(id string, f FileOutputFunc) FSSyncTarget {
+	return &fsSyncTarget{
+		id: id,
+		f:  f,
+	}
+}
+
+func WithFSSyncDir(id string, outdir string) FSSyncTarget {
+	return &fsSyncTarget{
+		id:     id,
 		outdir: outdir,
 	}
-	return p
 }
 
-// NewFSSyncTarget allows writing into an io.WriteCloser
-func NewFSSyncTarget(f func(map[string]string) (io.WriteCloser, error)) session.Attachable {
-	p := &fsSyncTarget{
-		f: f,
+func NewFSSyncTarget(targets ...FSSyncTarget) session.Attachable {
+	fs := make(map[string]FileOutputFunc)
+	outdirs := make(map[string]string)
+	defaultID := ""
+	for _, t := range targets {
+		t := t.target()
+		if t.f != nil {
+			fs[t.id] = t.f
+		}
+		if t.outdir != "" {
+			outdirs[t.id] = t.outdir
+		}
+		if defaultID == "" {
+			defaultID = t.id
+		}
+	}
+	return &fsSyncAttachable{
+		fs:        fs,
+		outdirs:   outdirs,
+		defaultID: defaultID,
 	}
-	return p
 }
 
-type fsSyncTarget struct {
-	outdir string
-	f      func(map[string]string) (io.WriteCloser, error)
+type fsSyncAttachable struct {
+	fs      map[string]FileOutputFunc
+	outdirs map[string]string
+
+	defaultID string
 }
 
-func (sp *fsSyncTarget) Register(server *grpc.Server) {
+func (sp *fsSyncAttachable) Register(server *grpc.Server) {
 	RegisterFileSendServer(server, sp)
 }
 
-func (sp *fsSyncTarget) DiffCopy(stream FileSend_DiffCopyServer) (err error) {
-	if sp.outdir != "" {
-		return syncTargetDiffCopy(stream, sp.outdir)
+func (sp *fsSyncAttachable) chooser(ctx context.Context) string {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return sp.defaultID
+	}
+	values := md[keyExporterID]
+	if len(values) == 0 {
+		return sp.defaultID
 	}
+	id := values[0]
+	return id
+}
 
-	if sp.f == nil {
-		return errors.New("empty outfile and outdir")
+func (sp *fsSyncAttachable) DiffCopy(stream FileSend_DiffCopyServer) (err error) {
+	id := sp.chooser(stream.Context())
+	if outdir, ok := sp.outdirs[id]; ok {
+		return syncTargetDiffCopy(stream, outdir)
 	}
+	f, ok := sp.fs[id]
+	if !ok {
+		return errors.Errorf("exporter %q not found", id)
```

Comment:
Given that hitting this is indicative of a buildkit internal error, I'd rather have the id printed here.

This would mean that somehow buildkit has requested to export to an exporter that the client never created.